src/python/grpcio_tests/tests_py3_only/interop/xds_interop_client.py

```diff
@@ -294,18 +295,17 @@ def _run_single_channel(config: _ChannelConfiguration) -> None:
                     continue
                 else:
                     duration_per_query = 1.0 / float(config.qps)
-            request_id = None
-            with _global_lock:
-                request_id = _global_rpc_id
-                _global_rpc_id += 1
-                _global_rpcs_started[config.method] += 1
-            start = time.time()
-            end = start + duration_per_query
-            with config.condition:
+                request_id = None
+                with _global_lock:
+                    request_id = _global_rpc_id
+                    _global_rpc_id += 1
+                    _global_rpcs_started[config.method] += 1
+                start = time.time()
+                end = start + duration_per_query
                 _start_rpc(config.method, config.metadata, request_id, stub,
                            float(config.rpc_timeout_sec), futures)
-            with config.condition:
-                _remove_completed_rpcs(futures, config.print_response)
+                print_response = config.print_response
+            _remove_completed_rpcs(futures, config.print_response)
```

Comment:
In this PR, we are making sure the creation of RPCs won't race with the configuration updates. Have we considered in-flight RPCs? In `_remove_completed_rpcs`, we are treating the RPCs with old spec and the RPCs with new spec the same way. The tear might still be a problem for slow-responding RPCs.

There is an alternative of cleaning the in-flight RPCs before making configuration changes. WDYT?

(In the test cases this PR add, adding slow-responding RPCs is not that easy, since it requires server-side changes. I'm okay if we want to defer that test case till future.)

Reply:
> In _remove_completed_rpcs, we are treating the RPCs with old spec and the RPCs with new spec the same way. The tear might still be a problem for slow-responding RPCs.
> There is an alternative of cleaning the in-flight RPCs before making configuration changes. WDYT?

I considered this as a potential cause of the problem and actually implemented the cancellation you're suggesting as in an attempt to fix the bug. But in practice, this is not a problem because all RPCs are created with a timeout defined by the test harness. If we cancel RPCs, then we'll end up with another bad situation -- RPCs with cancellation status codes in the accumulated stats.

If we decide to add this cancellation behavior, it should be added to the spec and to all other test clients.

Reply:
Good point about the cancellation. How about delay the configuration until the futures are cleared? In that case, we won't have unwanted status codes in the stats.

(With current design, the same configuration tear problem might apply to other languages as well.)

Reply:
I think this is out of the scope of this PR and the bug that I'm trying to solve here, so I've created [a tracking issue](https://github.com/grpc/grpc/issues/29442) for this discussion.